storage/src/commit_log/mod.rs

```diff
@@ -0,0 +1,430 @@
+// Copyright (c) SimpleStaking and Tezedge Contributors
+// SPDX-License-Identifier: MIT
+
+mod compression;
+
+use failure::Fail;
+use serde::{Deserialize, Serialize};
+
+use crate::persistent::codec::{Decoder, Encoder, SchemaError};
+use crate::persistent::schema::{CommitLogDescriptor, CommitLogSchema};
+use crate::persistent::BincodeEncoded;
+use std::{fmt, io};
+
+use std::path::{Path, PathBuf};
+use std::sync::{Arc, RwLock};
+
+use crate::commit_log::compression::{zstd_compress, zstd_decompress};
+use std::collections::HashMap;
+use std::fs::{File, OpenOptions};
+use std::io::{BufReader, BufWriter, Read, Seek, SeekFrom, Write};
+
+pub type CommitLogRef = Arc<RwLock<CommitLog>>;
+
+const DATA_FILE_NAME: &str = "table.data";
+
+pub struct CommitLog {
+    data_file: File,
+    data_file_path: PathBuf,
+    use_compression: bool,
+}
+
+impl CommitLog {
+    pub fn new<P: AsRef<Path>>(dir: P, use_compression: bool) -> Result<Self, CommitLogError> {
+        if !dir.as_ref().exists() {
+            std::fs::create_dir_all(dir.as_ref())?;
+        }
+        let mut data_file_path = PathBuf::new();
+        data_file_path.push(dir.as_ref());
+        data_file_path.push(DATA_FILE_NAME);
+        let data_file = OpenOptions::new()
+            .create(true)
+            .write(true)
+            .read(true)
+            .open(data_file_path.clone())?;
+        Ok(Self {
+            data_file,
+            data_file_path,
+            use_compression,
+        })
+    }
+    pub fn append_msg<B: AsRef<[u8]>>(
+        &mut self,
+        payload: B,
+    ) -> Result<(u64, usize), CommitLogError> {
+        let mut writer = BufWriter::new(&mut self.data_file);
+        let offset = writer.seek(SeekFrom::End(0))?;
+        let buf_size = if self.use_compression {
+            let mut compressed_payload = Vec::new();
+            zstd_compress(payload, &mut compressed_payload)?;
+            writer.write_all(compressed_payload.as_slice())?;
+            compressed_payload.len()
+        } else {
+            writer.write_all(payload.as_ref())?;
+            payload.as_ref().len()
+        };
+        writer.flush()?;
+        Ok((offset, buf_size))
+    }
+
+    pub fn read(&self, offset: u64, buf_size: usize) -> Result<Vec<u8>, CommitLogError> {
+        let mut buf = vec![0_u8; buf_size];
+        let mut reader = BufReader::new(File::open(self.data_file_path.as_path())?);
```

Comment:
@mambisi 
do we need to File::open here? couldnt we just use `data_file: File,` ?

and another thing: `self.data_file_path.as_path()` could it be retrieved from `data_file: File,` ? if so, then we dont need `self.data_file_path`

Reply:
https://doc.rust-lang.org/std/fs/struct.File.html,
std::fs:File has not `as_path()` method

Reply:
File might be auto sync when you try to read it without opening it, for the testing this was best to get consistent results in reading a file